File: backend/app/models/meal_share.py
from __future__ import annotations

import uuid

from sqlalchemy import Boolean, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, CreatedAt, UuidPK


class MealShare(Base):
    """Post-checkout 'split the cost' link.

    Created after an order is paid: we mint a fixed-amount bunq.me URL via
    the existing MCP payment infrastructure (per-person share, not the full
    total) and store it here so the user can re-open the same link later
    instead of generating a new one each time."""

    __tablename__ = "meal_shares"

    id: Mapped[UuidPK]
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("orders.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    participant_count: Mapped[int] = mapped_column(Integer, nullable=False)
    include_self: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    per_person_eur: Mapped[float] = mapped_column(Float, nullable=False)
    total_eur: Mapped[float] = mapped_column(Float, nullable=False)
    bunq_request_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    share_url: Mapped[str] = mapped_column(String(2048), nullable=False)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="open")
    created_at: Mapped[CreatedAt]

File: ios/Sources/Models/MealShare.swift
import Foundation

/// Wire shape for `POST /orders/{order_id}/share-cost` and
/// `GET /orders/{order_id}/share-cost`. The bunq.me URL is fixed-amount per
/// person — friends pay their share via iDEAL/card/bank.
struct MealShare: Codable, Identifiable, Hashable {
    let id: String
    let orderId: String
    let participantCount: Int
    let includeSelf: Bool
    let perPersonEur: Double
    let totalEur: Double
    let shareURL: String
    let bunqRequestId: String?
    let status: String          // "open" | "closed"
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case orderId          = "order_id"
        case participantCount = "participant_count"
        case includeSelf      = "include_self"
        case perPersonEur     = "per_person_eur"
        case totalEur         = "total_eur"
        case shareURL         = "share_url"
        case bunqRequestId    = "bunq_request_id"
        case status
        case createdAt        = "created_at"
    }

    /// Total people the bill is divided across — including the owner if
    /// `includeSelf` is true.
    var divisor: Int { participantCount + (includeSelf ? 1 : 0) }
}
